File: rocketmq-go/consumer/pull_consumer.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package consumer

import (
	"fmt"
	"github.com/apache/incubator-rocketmq-externals/rocketmq-go/model"
	"github.com/apache/incubator-rocketmq-externals/rocketmq-go/model/message"
	"github.com/apache/incubator-rocketmq-externals/rocketmq-go/remoting"
	"github.com/apache/incubator-rocketmq-externals/rocketmq-go/service"
	"github.com/golang/glog"
	"time"
)

type DefaultMQPullConsumer struct {
	consumerStartTimestamp time.Time
	rpcHook                remoting.RPCHook
	consumeMessageHookList []model.ConsumerHook // TODO consumeMessageHook
	filterMessageHookList  []model.ConsumerHook // TODO filterMessageHook
	status                 ServiceStatus
	mqClient               *service.MQClient
	//clientAPI              *service.MQClientAPI
	offsetStore service.OffsetStore
	rebalance   service.Rebalance
	cfg         RocketMqConsumerConfig
	wrapper     pullAPIWrapper
}

func NewDefaultMQPullConsumer(cfg RocketMqConsumerConfig, hook remoting.RPCHook) DefaultMQPullConsumer {
	// TODO
	return DefaultMQPullConsumer{
		rpcHook:   hook,
		rebalance: service.PullMessageRebalance{},
		cfg:       cfg}
}

func (dpc *DefaultMQPullConsumer) RegisterConsumeMessageHook(hook model.ConsumerHook) {
	// TODO optimize
	dpc.consumeMessageHookList[len(dpc.consumeMessageHookList)] = hook
	glog.Infof("register consumeMessageHook Hook, %s", hook.Name())
}

func (dpc *DefaultMQPullConsumer) CreateTopic(key, topic string, queueNum, topicSysFlag int) error {
	dpc.makeSureStatusOK()
	// TODO
	//return dpc.api.CreateTopic(key, topic, queueNum, topicSysFlag)
	return nil
}

func (dpc *DefaultMQPullConsumer) makeSureStatusOK() {
	if dpc.status != Running {
		glog.Fatal(model.NewMQClientError(-1, fmt.Sprintf("The consumer service state not OK, %s", dpc.status)))
	}
}

func (dpc *DefaultMQPullConsumer) FetchConsumeOffset(mq *message.MessageQueue, readType service.ReadOffsetType) (int64, error) {
	dpc.makeSureStatusOK()
	return dpc.offsetStore.ReadOffset(mq, readType)
}

func (dpc *DefaultMQPullConsumer) FetchMessageQueuesInBalance(topic string) ([]*message.MessageQueue, error) {
	dpc.makeSureStatusOK()
	// TODO
	return nil, nil
}

func (dpc *DefaultMQPullConsumer) SendMessageBack(msg message.MessageExt, delayLevel int,
	brokerName, consumerGroup string) error {
	// TODO with pullAPIWrapper

	return nil
}

func (dpc *DefaultMQPullConsumer) Pull(mq *message.MessageQueue, subExp string, offset int64,
	maxNum int, timeout time.Duration) model.PullResult {
	return dpc.pullSync(mq, subExp, offset, maxNum, false, timeout)
}

func (dpc *DefaultMQPullConsumer) pullSync(mq *message.MessageQueue, subExp string, offset int64,
	maxNum int, block bool, timeout time.Duration) model.PullResult {
	dpc.makeSureStatusOK()

	if mq == nil {
		glog.Fatal("MessageQueue is nil!")
	}

	if offset < 0 {
		glog.Fatal("offset < 0!")
	}

	if maxNum <= 0 {
		glog.Fatalf("maxNum <= 0")
	}

	dpc.subscriptionAutomatically(mq.Topic())

	sysFlag := BuildSysFlag(false, block, true, false)
	subscriptionData, err := model.BuildSubscriptionData(mq.Topic(), subExp)

	if err != nil {
		// TODO log
	}

	// TODO refactor api
	//long timeoutMillis = block ? this.defaultMQPullConsumer.getConsumerTimeoutMillisWhenSuspend() : timeout;
	pullResult, err := dpc.wrapper.pullKernel(mq,
		subscriptionData.SubString(),
		int64(0),
		offset,
		0,
		maxNum,
		sysFlag,
		dpc.cfg.brokerSuspendMaxTime,
		timeout,
		remoting.Sync,
		nil)

	if err != nil {
		glog.Errorf("PullMessage Error! Topic %s, BrokerName: %s", mq.Topic(), mq.BrokerName())
	}
	dpc.wrapper.processPullResult(mq, pullResult, subscriptionData)

	if dpc.consumeMessageHookList != nil && len(dpc.consumeMessageHookList) > 0 {
		// TODO
	}
	return pullResult
}

func (dpc *DefaultMQPullConsumer) subscriptionAutomatically(topic string) {
	if _, found := dpc.rebalance.SubscriptionInner()[topic]; !found {
		subscriptionData, err := model.BuildSubscriptionData(topic, model.SubscribeAll)
		if err != nil {
			// TODO log, but can be ignored
		}
		dpc.rebalance.SubscriptionInner()[topic] = subscriptionData // TODO optimize
	}
}

func (dpc *DefaultMQPullConsumer) UnSubscribe(topic string) {
	delete(dpc.rebalance.SubscriptionInner(), topic) // TODO optimize
}

func (dpc *DefaultMQPullConsumer) GroupName() string

func (dpc *DefaultMQPullConsumer) executeHookBefore(ctx model.ConsumeMessageContext) {
	for _, hook := range dpc.consumeMessageHookList {
		hook.DoBefore()
	}
}

func (dpc *DefaultMQPullConsumer) executeHookAfter(ctx model.ConsumeMessageContext) {
	for _, hook := range dpc.consumeMessageHookList {
		hook.DoAfter()
	}
}

func (dpc *DefaultMQPullConsumer) MessageMode() MessageModel

func (dpc *DefaultMQPullConsumer) ConsumeType() ConsumeType {
	return ConsumeActively
}

func (dpc *DefaultMQPullConsumer) ConsumeFromWhere() ConsumeFromWhere {
	return ConsumeFromLastOffset
}

func (dpc *DefaultMQPullConsumer) Subscriptions() []*model.SubscriptionData {
	var result []*model.SubscriptionData
	// TODO
	return result
}

func (dpc *DefaultMQPullConsumer) DoRebalance() {
	if dpc.rebalance != nil {
		dpc.rebalance.DoRebalance(false)
	} else {
		glog.Fatalf("Rebalance Service is nil!")
	}
}

func (dpc *DefaultMQPullConsumer) PersistConsumerOffset() {
	dpc.makeSureStatusOK()

	allocateMQ := dpc.rebalance.ProcessQueueTable()

	var mqs []*message.MessageQueue

	var index = 0 // TODO optimize
	for k := range allocateMQ {
		mqs[index] = &k
		index++
	}
	dpc.offsetStore.PersistAll(mqs)
}

func (dpc *DefaultMQPullConsumer) UpdateTopicSubscribeInfo(topic string, info []*message.MessageQueue) {
	subTable := dpc.rebalance.SubscriptionInner() // TODO optimize
	if subTable != nil {
		dpc.rebalance.TopicSubscribeInfoTable()[topic] = info
	}
}

func (dpc *DefaultMQPullConsumer) IsSubscribeTopicNeedUpdate(topic string) bool {
	subTable := dpc.rebalance.SubscriptionInner()
	if subTable != nil {
		_, found := subTable[topic]
		if found {
			_, found = dpc.rebalance.TopicSubscribeInfoTable()[topic]
		}
		return found
	}
	return false
}

func (dpc *DefaultMQPullConsumer) UnitMode() bool {
	// TODO
	return false
}

// TODO consider the method position
func (dpc *DefaultMQPullConsumer) ConsumerRunningInfo() runningInfo

func (dpc *DefaultMQPullConsumer) PullAsync(mq *message.MessageQueue, subExp string, offset int64,
	maxNum int, timeout time.Duration, callback model.PullCallback) {
	dpc.pullAsync(mq, subExp, offset, maxNum, false, timeout, callback)
}

func (dpc *DefaultMQPullConsumer) pullAsync(mq *message.MessageQueue, subExp string, offset int64,
	maxNum int, block bool, timeout time.Duration, callback model.PullCallback) {
	dpc.makeSureStatusOK()

	if mq == nil {
		glog.Fatal("MessageQueue is nil!")
	}

	if offset < 0 {
		glog.Fatal("offset < 0!")
	}

	if maxNum <= 0 {
		glog.Fatal("maxNum <= 0")
	}

	if callback == nil {
		glog.Fatal("PullCallback is nil!")
	}

	dpc.subscriptionAutomatically(mq.Topic())

	sysFlag := BuildSysFlag(false, block, true, false)

	subscriptionData, err := model.BuildSubscriptionData(mq.Topic(), subExp)
	// TODO refactor api
	if err != nil {
		glog.Fatal("parse subscription error!")
	}

	// TODO
	//long timeoutMillis = block ? this.defaultMQPullConsumer.getConsumerTimeoutMillisWhenSuspend() : timeout;
	dpc.wrapper.pullKernel(mq,
		subscriptionData.SubString(),
		int64(0),
		offset,
		0,
		maxNum,
		sysFlag,
		dpc.cfg.brokerSuspendMaxTime,
		timeout,
		remoting.Async,
		callback) // TODO optimize: use advanced function
}

func (dpc *DefaultMQPullConsumer) PullBlockIfNotFound(mq *message.MessageQueue, subExp string, offset int64, maxNum int) model.PullResult {
	return dpc.pullSync(mq, subExp, offset, maxNum, true, time.Second) // TODO optimize time
}

func (dpc *DefaultMQPullConsumer) PullBlockIfNotFoundAsync(mq *message.MessageQueue, subExp string, offset int64,
	maxNum int, callback model.PullCallback) {
	dpc.pullAsync(mq, subExp, offset, maxNum, true, time.Second, callback) // TODO optimize time
}

func (dpc *DefaultMQPullConsumer) Start() {
	switch dpc.status {
	case CreateJust:
		dpc.status = StartFailed

		dpc.checkConfig()
		dpc.copySubscription()
		dpc.wrapper.api.Start()
		// TODO
	default:
		glog.Fatalf("The PullConsumer service status not OK, maybe started once. STATUS: %s", dpc.status)
	}
}

func (dpc *DefaultMQPullConsumer) Shutdown() {
	switch dpc.status {
	case Running:
		dpc.PersistConsumerOffset()
		dpc.wrapper.api.Shutdown()
		dpc.mqClient.RemoveConsumer(dpc.cfg.consumerGroup)
		glog.Infof("The consumer [%s] shutdown successfully.", dpc.cfg.consumerGroup)
		dpc.status = ShutdownAlready
	default:
		glog.Warning("The Consumer no Running!")
		break
	}
}

func (dpc *DefaultMQPullConsumer) checkConfig() {
	// TODO
}

func (dpc *DefaultMQPullConsumer) copySubscription() {
	// TODO
}

File: rocketmq-client-php/src/pull_consumer.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#include "pull_consumer.h"
#include "session_credentials.h"

void PullConsumer::__construct(Php::Parameters &params){
    std::string groupName = params[0];
    this->consumer = new rocketmq::DefaultMQPullConsumer(groupName);
}

void PullConsumer::setGroup(Php::Parameters &params){
    std::string groupName = params[0];
    this->consumer->setGroupName(groupName);
}

void PullConsumer::start(){
    this->consumer->start();
}

Php::Value PullConsumer::getQueues(){
    Php::Array result;
    int idx = 0;
    this->consumer->fetchSubscribeMessageQueues(this->topicName, this->mqs);
    std::vector<rocketmq::MQMessageQueue>::iterator iter = mqs.begin();
    for (; iter != mqs.end(); ++iter) {
        rocketmq::MQMessageQueue mq = (*iter);
        result[idx++] = Php::Object(MESSAGE_QUEUE_CLASS_NAME , new MessageQueue(mq)); 
    }
    return result;
}

Php::Value PullConsumer::getNamesrvDomain(){
    return this->consumer->getNamesrvDomain();
}

void PullConsumer::setNamesrvDomain(Php::Parameters &param){
    std::string namesrv_domain = param[0];
    this->consumer->setNamesrvDomain(namesrv_domain);
}

Php::Value PullConsumer::getNamesrvAddr(){
    return this->consumer->getNamesrvAddr();
}

void PullConsumer::setNamesrvAddr(Php::Parameters &param){
    std::string namesrv_addr = param[0];
    this->consumer->setNamesrvAddr(namesrv_addr);
}

void PullConsumer::setInstanceName(Php::Parameters &param){
    std::string instanceName = param[0];
    this->consumer->setInstanceName(instanceName);
}

void PullConsumer::setTopic(Php::Parameters &param){
    std::string topic = param[0];
    this->topicName= topic;
}

// pull( MessageQueue mq, string subExpression, int offset, int maxNums)
Php::Value PullConsumer::pull(Php::Parameters &param){
    Php::Value mq = param[0];
    std::string subExpression = param[1];
    int64_t offset = param[2];
    int64_t maxNums = param[3];
    MessageQueue* messageQueue = (MessageQueue*)mq.implementation();
    rocketmq::PullResult result = this->consumer->pull(messageQueue->getInstance(), subExpression, offset, maxNums);
    PullResult *pullResult = new PullResult(result);
    Php::Value pv(Php::Object(PULL_RESULT_CLASS_NAME, pullResult));
    return pv;
}

/*
   void PullConsumer::persistConsumerOffset(){
   this->consumer->persistConsumerOffset();
   }

   void PullConsumer::persistConsumerOffsetByResetOffset(){
   this->consumer->persistConsumerOffsetByResetOffset();
   }
   */

Php::Value PullConsumer::pullBlockIfNotFound(Php::Parameters &param){
    Php::Value mq = param[0];
    std::string subExpression = param[1];
    int64_t offset = param[2];
    int64_t maxNums = param[3];
    MessageQueue* messageQueue = (MessageQueue*)mq.implementation();
    rocketmq::PullResult result = this->consumer->pullBlockIfNotFound(messageQueue->getInstance(), subExpression, offset, maxNums);
    PullResult *pullResult = new PullResult(result);
    Php::Value pv(Php::Object(PULL_RESULT_CLASS_NAME, pullResult));
    return pv;
}

void PullConsumer::setSessionCredentials(Php::Parameters &param){
    std::string accessKey = param[0];
    std::string secretKey = param[1];
    std::string authChannel = param[2];

    this->consumer->setSessionCredentials(accessKey, secretKey, authChannel);
}

Php::Value PullConsumer::getSessionCredentials(){
    rocketmq::SessionCredentials sc = this->consumer->getSessionCredentials();
    SessionCredentials *sessionCredentials = new SessionCredentials(&sc);
    Php::Value pv(Php::Object(SESSION_CREDENTIALS_CLASS_NAME , sessionCredentials));
    return pv;
}


void PullConsumer::updateConsumeOffset(Php::Parameters &params){
    Php::Value mq = params[0];
    int64_t offset = params[1];

    MessageQueue* messageQueue = (MessageQueue*)mq.implementation();
    this->consumer->updateConsumeOffset(messageQueue->getInstance(), offset);
}

void PullConsumer::removeConsumeOffset(Php::Parameters &params){
    Php::Value mq = params[0];
    MessageQueue* messageQueue = (MessageQueue*)mq.implementation();
    this->consumer->removeConsumeOffset(messageQueue->getInstance());
}

Php::Value PullConsumer::fetchConsumeOffset(Php::Parameters &params){
    Php::Value mq = params[0];
    bool fromStore = params[1];
    MessageQueue* messageQueue = (MessageQueue*)mq.implementation();
    return (int64_t)this->consumer->fetchConsumeOffset(messageQueue->getInstance(), fromStore);
}

Php::Value PullConsumer::getMessageModel(){
    return this->consumer->getMessageModel();
}
void PullConsumer::setMessageModel(Php::Parameters &params){
    this->consumer->setMessageModel(rocketmq::MessageModel((int)params[0]));
}

// void setTcpTransportPullThreadNum(int num);
void PullConsumer::setTcpTransportPullThreadNum(Php::Parameters &param){
    this->consumer->setTcpTransportPullThreadNum((int64_t)param[0]);
}

// const int getTcpTransportPullThreadNum() const;
Php::Value PullConsumer::getTcpTransportPullThreadNum(){
    return (int64_t)this->consumer->getTcpTransportPullThreadNum();
}

// void setTcpTransportConnectTimeout(uint64_t timeout);  // ms
void PullConsumer::setTcpTransportConnectTimeout(Php::Parameters &param){
    this->consumer->setTcpTransportConnectTimeout((int64_t)param[0]);
}
// const uint64_t getTcpTransportConnectTimeout() const;
Php::Value PullConsumer::getTcpTransportConnectTimeout(){
    return (int64_t)this->consumer->getTcpTransportConnectTimeout();
}

// void setTcpTransportTryLockTimeout(uint64_t timeout);  // ms
void PullConsumer::setTcpTransportTryLockTimeout(Php::Parameters &param){
    this->consumer->setTcpTransportTryLockTimeout((int64_t)param[0]);
}

// const uint64_t getTcpTransportConnectTimeout() const;
Php::Value PullConsumer::getTcpTransportTryLockTimeout(){
    return (int64_t)this->consumer->getTcpTransportTryLockTimeout();
}

//void setUnitName(std::string unitName);
void PullConsumer::setUnitName(Php::Parameters &param){
    this->consumer->setUnitName(param[0]);
}
//const std::string& getUnitName();
Php::Value PullConsumer::getUnitName(){
    return this->consumer->getUnitName();
}

void PullConsumer::setLogLevel(Php::Parameters &param){
    this->consumer->setLogLevel(rocketmq::elogLevel((int)param[0]));
}

Php::Value PullConsumer::getLogLevel(){
    return this->consumer->getLogLevel();
}

void PullConsumer::setLogFileSizeAndNum(Php::Parameters &param){
    this->consumer->setLogFileSizeAndNum(param[0], param[1]);
}



void registerPullConsumer(Php::Namespace &rocketMQNamespace){
    Php::Class<PullConsumer> pullConsumer("PullConsumer");
    pullConsumer.method<&PullConsumer::__construct>("__construct", { Php::ByVal("groupName", Php::Type::String), });
    pullConsumer.method<&PullConsumer::setInstanceName>("setInstanceName", { Php::ByVal("instance", Php::Type::String), });
    pullConsumer.method<&PullConsumer::setTopic>("setTopic", { Php::ByVal("topic", Php::Type::String), });
    pullConsumer.method<&PullConsumer::start>("start");
    pullConsumer.method<&PullConsumer::getQueues>("getQueues");

    pullConsumer.method<&PullConsumer::setNamesrvAddr>("setNamesrvAddr", { Php::ByVal("namesrvAddr", Php::Type::String), });
    pullConsumer.method<&PullConsumer::getNamesrvAddr>("getNamesrvAddr");

    pullConsumer.method<&PullConsumer::setNamesrvDomain>("setNamesrvDomain", { Php::ByVal("nameserver", Php::Type::String), });
    pullConsumer.method<&PullConsumer::getNamesrvDomain>("getNamesrvDomain");

    pullConsumer.method<&PullConsumer::setGroup>("setGroup", { Php::ByVal("group", Php::Type::String), });
    pullConsumer.method<&PullConsumer::pull>("pull", {
            Php::ByVal("mq", MESSAGE_QUEUE_CLASS_NAME),
            Php::ByVal("subExpression", Php::Type::String),
            Php::ByVal("offset", Php::Type::Numeric),
            Php::ByVal("maxNums", Php::Type::Numeric),
            });

    pullConsumer.method<&PullConsumer::pullBlockIfNotFound>("pullBlockIfNotFound", {
            Php::ByVal("mq", MESSAGE_QUEUE_CLASS_NAME),
            Php::ByVal("subExpression", Php::Type::String),
            Php::ByVal("offset", Php::Type::Numeric),
            Php::ByVal("maxNums", Php::Type::Numeric),
            });

    pullConsumer.method<&PullConsumer::setSessionCredentials>("setSessionCredentials", {
            Php::ByVal("accessKey", Php::Type::String),
            Php::ByVal("secretKey", Php::Type::String),
            Php::ByVal("authChannel", Php::Type::String),
            });
    pullConsumer.method<&PullConsumer::getSessionCredentials>("getSessionCredentials");
    pullConsumer.method<&PullConsumer::updateConsumeOffset>("updateConsumeOffset", {
            Php::ByVal("mq", MESSAGE_QUEUE_CLASS_NAME),
            Php::ByVal("offset", Php::Type::Numeric),
            });
    pullConsumer.method<&PullConsumer::removeConsumeOffset>("removeConsumeOffset", { Php::ByVal("mq", MESSAGE_QUEUE_CLASS_NAME), });
    pullConsumer.method<&PullConsumer::fetchConsumeOffset>("fetchConsumeOffset", {
            Php::ByVal("mq", MESSAGE_QUEUE_CLASS_NAME),
            Php::ByVal("fromStore", Php::Type::Bool),
            });
    pullConsumer.method<&PullConsumer::getMessageModel>("getMessageModel");
    pullConsumer.method<&PullConsumer::setMessageModel>("setMessageModel", { Php::ByVal("messageModel", Php::Type::Numeric), });

    pullConsumer.method<&PullConsumer::getTcpTransportTryLockTimeout>("getTcpTransportTryLockTimeout");
    pullConsumer.method<&PullConsumer::setTcpTransportTryLockTimeout>("setTcpTransportTryLockTimeout",{ Php::ByVal("timeout", Php::Type::Numeric), });

    pullConsumer.method<&PullConsumer::getTcpTransportConnectTimeout>("getTcpTransportConnectTimeout");
    pullConsumer.method<&PullConsumer::setTcpTransportConnectTimeout>("setTcpTransportConnectTimeout", {Php::ByVal("timeout", Php::Type::Numeric), });

    pullConsumer.method<&PullConsumer::getTcpTransportPullThreadNum>("getTcpTransportPullThreadNum", {Php::ByVal("threadNum", Php::Type::Numeric), });
    pullConsumer.method<&PullConsumer::setTcpTransportPullThreadNum>("setTcpTransportPullThreadNum", {Php::ByVal("threadNum", Php::Type::Numeric), });

    pullConsumer.method<&PullConsumer::getUnitName>("getUnitName");
    pullConsumer.method<&PullConsumer::setUnitName>("setUnitName", {Php::ByVal("unitName", Php::Type::String),});

    pullConsumer.method<&PullConsumer::setLogLevel>("setLogLevel", {Php::ByVal("inputLevel", Php::Type::Numeric),});
    pullConsumer.method<&PullConsumer::getLogLevel>("getLogLevel");
    pullConsumer.method<&PullConsumer::setLogFileSizeAndNum>("setLogFileSizeAndNum", {Php::ByVal("fileNum", Php::Type::Numeric),Php::ByVal("perFileSize", Php::Type::Numeric),});



    rocketMQNamespace.add(pullConsumer);
}
